File: builder.py
class Computadora:
    def __init__(self):
        self.partes = []

    def agregar_parte(self, parte):
        self.partes.append(parte)

    def mostrar(self):
        return f"Computadora con: {', '.join(self.partes)}"

class ConstructorComputadora:
    def __init__(self):
        self.pc = Computadora()

    def agregar_cpu(self):
        self.pc.agregar_parte("CPU Intel i7")

    def agregar_ram(self):
        self.pc.agregar_parte("16GB RAM")

    def agregar_gpu(self):
        self.pc.agregar_parte("NVIDIA RTX 3060")

    def obtener_computadora(self):
        return self.pc


class Director:
    def construir_gaming_pc(self, constructor):
        constructor.agregar_cpu()
        constructor.agregar_ram()
        constructor.agregar_gpu()

constructor = ConstructorComputadora()
director = Director()
director.construir_gaming_pc(constructor)
pc = constructor.obtener_computadora()
print(pc.mostrar())

File: java/builder.java
import java.util.ArrayList;
import java.util.List;

class Computadora {
    private List<String> partes = new ArrayList<>();

    public void agregarParte(String parte) {
        partes.add(parte);
    }

    public String mostrar() {
        return "Computadora con: " + String.join(", ", partes);
    }
}

class ConstructorComputadora {
    private Computadora pc;

    public ConstructorComputadora() {
        this.pc = new Computadora();
    }

    public void agregarCPU() {
        pc.agregarParte("CPU Intel i7");
    }

    public void agregarRAM() {
        pc.agregarParte("16GB RAM");
    }

    public void agregarGPU() {
        pc.agregarParte("NVIDIA RTX 3060");
    }

    public Computadora obtenerComputadora() {
        return pc;
    }
}


class Director {
    public void construirGamingPC(ConstructorComputadora constructor) {
        constructor.agregarCPU();
        constructor.agregarRAM();
        constructor.agregarGPU();
    }
}


public class builder{
    public static void main(String[] args) {
        ConstructorComputadora constructor = new ConstructorComputadora();
        Director director = new Director();

        director.construirGamingPC(constructor);

        Computadora pc = constructor.obtenerComputadora();
        System.out.println(pc.mostrar());
    }
}
